dms/models/dms_file.py

```diff
@@ -0,0 +1,629 @@
+        self.ensure_one()
+        default = dict(default or [])
+        if "directory_id" in default:
+            model = self.env["dms.directory"]
+            directory = model.browse(default["directory_id"])
+            names = directory.sudo().file_ids.mapped("name")
+        else:
+            names = self.sudo().directory_id.file_ids.mapped("name")
+        default.update({"name": file.unique_name(self.name, names, self.extension)})
+        return super(File, self).copy(default)
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        new_vals_list = []
+        for vals in vals_list:
+            if "attachment_id" not in vals:
+                vals = self._create_model_attachment(vals)
+            new_vals_list.append(vals)
+        return super(File, self).create(new_vals_list)
+
+    # ----------------------------------------------------------
+    # Locking fields and functions
+    # ----------------------------------------------------------
+
+    locked_by = fields.Many2one(comodel_name="res.users")
+
+    is_locked = fields.Boolean(compute="_compute_locked", string="Locked")
+
+    is_lock_editor = fields.Boolean(compute="_compute_locked", string="Editor")
+
+    # ----------------------------------------------------------
+    # Locking
+    # ----------------------------------------------------------
+
+    def lock(self):
+        self.write({"locked_by": self.env.uid})
+
+    def unlock(self):
+        self.write({"locked_by": None})
+
+    # ----------------------------------------------------------
+    # Read, View
+    # ----------------------------------------------------------
+
+    @api.depends("locked_by")
+    def _compute_locked(self):
+        for record in self:
+            if record.locked_by.exists():
+                record.update(
+                    {
+                        "is_locked": True,
+                        "is_lock_editor": record.locked_by.id == record.env.uid,
+                    }
+                )
+            else:
+                record.update({"is_locked": False, "is_lock_editor": False})
+
+    def create_dms_file_from_attachments(self, attachment_ids=None, view_type="list"):
+        """Create the dms files from files.
+        :return: An action redirecting to dms.file tree view.
+        """
+
+        if attachment_ids is None:
+            attachment_ids = []
+
+        attachments = self.env["ir.attachment"].browse(attachment_ids)
+
+        if not attachments:
+            raise UserError(_("No attachment was provided"))
+
+        if any(
+            attachment.res_id or attachment.res_model != "dms.file"
+            for attachment in attachments
+        ):
+            raise UserError(_("Invalid attachments!"))
+
+        for attachment in attachments:
+            return {
+                "name": attachment.name,
+                "datas": attachment.datas,
+                "res_model": attachment.res_model,
+            }
```

Comment:
```suggestion
        return [{
                "name": attachment.name,
                "datas": attachment.datas,
                "res_model": attachment.res_model,
            } for attachment in attachments]
```

It is a bit weird that it would raise and not simply return an empty list if it's supposed to return a list.

Otherwise, replace the `attachment_ids` argument by `attachment_id` and simply return a dict.

In any case, this implementation always return only one attachment, which is quite wrong and should have been caught by some unit test. (let me know if you need some help).

Reply:
thanks, if you have time, you can look at the error of the pre-commit :)

Reply:
Nice. You can cherry-pick the commit https://github.com/len-foss/dms/commit/dd9994297fc42ef3bb76a26c0dee1bff555515fb and squash it. Note that I rebased on `origin/16.0` before (you get the pre-commit error locally if you rebase).